File: testdata/fixtures/ocaml/CWE-77.ml
(* CWE-77: OCaml Unix.system shell metachar [|;&$`] *)
let vuln1 user_input =
  (* ruleid: ocaml-77-shell *)
  let _ = Unix.system user_input in
  ()

let vuln2 cmd =
  (* ruleid: ocaml-77-shell *)
  let _ = Sys.command cmd in
  ()

let vuln3 arg =
  let cmd = arg in
  (* ruleid: ocaml-77-shell *)
  let _ = Unix.system cmd in
  ()

let safeAllowlist () =
  let allowlist = read_line () in
  if not (String.contains allowlist 'l') then () else
  (* ok: ocaml-77-shell *)
  let _ = Unix.system allowlist in
  ()

let safeFixed () =
  (* ok: ocaml-77-shell *)
  let _ = Unix.system "ls -la" in
  ()

File: testdata/fixtures/javascript/CWE-77.js
// CWE-77: Command injection via child_process.exec with shell metachar
const child_process = require('child_process');
const express = require('express');
const app = express();

app.get('/exec1', (req, res) => {
  const cmd = req.query.cmd;
  // ruleid: javascript-77-shell
  child_process.exec(cmd);
});

app.get('/exec2', (req, res) => {
  const input = req.params.input;
  // ruleid: javascript-77-shell
  child_process.execSync(input);
});

app.post('/exec3', (req, res) => {
  const bodyCmd = req.body.cmd;
  // ruleid: javascript-77-shell
  child_process.spawn(bodyCmd, { shell: true });
});

function safeArray(req, res) {
  const arg = req.query.arg;
  // ok: javascript-77-shell
  child_process.spawn("ls", [arg], { shell: false });
}

function safeFixed(req, res) {
  // ok: javascript-77-shell
  child_process.exec("ls -la");
}
